plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -1140,15 +1143,19 @@ public boolean downloadToFile(String urlString, Plugin plugin, @CheckForNull Fil
                     .setRetryHandler(new DefaultHttpRequestRetryHandler(maxRetries, true))
                     .build()) {
                 HttpClientContext context = HttpClientContext.create();
-                HttpHead httphead = new HttpHead(urlString);
-                try (CloseableHttpResponse ignored = httpclient.execute(httphead, context)) {
-                    HttpHost target = context.getTargetHost();
-                    List<URI> redirectLocations = context.getRedirectLocations();
-                    // Expected to be an absolute URI
-                    URI location = URIUtils.resolve(httphead.getURI(), target, redirectLocations);
-                    logVerbose(String.format("downloading %s from %s", plugin.getName(), location));
-                    FileUtils.copyURLToFile(location.toURL(), pluginFile);
-                } catch (URISyntaxException | IOException e) {
+                CredentialsProvider credentialsProvider = getCredentialsProvider();
+                if (credentialsProvider != null) {
+                    context.setCredentialsProvider(credentialsProvider);
+                }
+                HttpGet httpGet = new HttpGet(urlString);
+                try {
+                    httpclient.execute(httpGet, new FileDownloadResponseHandler(pluginFile), context);
+                    // get final URI (after all redirects)
+                    List<URI> locations = context.getRedirectLocations();
+                    if (locations != null) {
+                        logVerbose(String.format("Downloading %s from %s", plugin.getName(), locations.get(locations.size() - 1)));
```

Comment:
isn't this done after the request has been made?
```suggestion
                        logVerbose(String.format("Downloaded %s from %s", plugin.getName(), locations.get(locations.size() - 1)));
```

and shouldn't it be added to the catch clause in that case?

Reply:
Good catch, now logged in finally (so that it covers both success and failure): https://github.com/jenkinsci/plugin-installation-manager-tool/pull/292/commits/968d7c0d523e432d4e88f74536770ea9009c4233